torch/distributed/_spmd/data_parallel.py

```diff
@@ -0,0 +1,763 @@
+                            # activation must stay sharded
+                            act_spec, _ = batch_dim_analzer.get_batch_dim_shard_spec(
+                                arg, mesh
+                            )
+
+                            input_specs.append(act_spec)
+                        elif node_type == NodeType.PARAM:
+                            # param must be replicated
+                            input_specs.append(
+                                DTensorSpec(mesh=mesh, placements=[Replicate()])
+                            )
+                        else:
+                            raise RuntimeError(
+                                f"Expecting node with parameter and activation, but found {input_node_types}! "
+                            )
+                    # produce activation type sharding for output
+                    (
+                        output_spec,
+                        reduction_over_batch,
+                    ) = batch_dim_analzer.get_batch_dim_shard_spec(node, mesh)
+
+                    shard_strategy = PlacementStrategy(
+                        output_spec=output_spec, input_specs=input_specs
+                    )
+
+                    dp_strategy_map[node] = DataParallelStrategy(
+                        NodeType.ACT, [shard_strategy], reduction_over_batch
+                    )
+                elif NodeType.GRAD in input_node_types:
+                    # param/state + grad, build up acceptable strategy
+                    # the strategy should be the same for all the inputs/outputs
+                    # TODO: optimizer parts should follow the dtensor prop logic
+                    # to support more general cases that allows optimizer states
+                    # to have different shardings compare to the params
+                    replica_strategy = _gen_replicate_strategy(mesh)
+                    shard_strategy = _gen_shard_strategy(mesh, batch_dim)
+                    output_node_type = NodeType.PARAM
+                    for arg_node_type in input_node_types:
+                        if arg_node_type != NodeType.GRAD:
+                            output_node_type = arg_node_type
+                            break
+                    assert (
+                        output_node_type == NodeType.STATE
+                        or output_node_type == NodeType.PARAM
+                    ), "Expecting output node type to be either state or param! "
+
+                    dp_strategy_map[node] = DataParallelStrategy(
+                        output_node_type, [replica_strategy, shard_strategy]
+                    )
+                elif NodeType.STATE in input_node_types:
+                    # either param + state or state + state
+                    replica_strategy = _gen_replicate_strategy(mesh)
+                    shard_strategy = _gen_shard_strategy(mesh, batch_dim)
+                    output_node_type = (
+                        NodeType.PARAM
```

Comment:
is this guaranteed to be true, i.e., output is guaranteed to updated param as long as param is in the args?